### PR TITLE
ENH: added API stability decorators

### PR DESCRIPTION
**Pull request by gregcaporaso**

### CODE REVIEW

File: skbio/sequence/_protein.py [hunk 1]
```diff
@@ -102,16 +103,19 @@ def _stop_codes(cls):
 
     @classproperty
     @overrides(IUPACSequence)
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_protein.py]: bump this on top of overrides

File: skbio/sequence/_protein.py [hunk 2]
```diff
@@ -102,16 +103,19 @@ def _stop_codes(cls):
...
     def alphabet(cls):
         return super(Protein, cls).alphabet | cls.stop_chars
 
     @classproperty
     @overrides(IUPACSequence)
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_protein.py]: bump this on top of overrides

File: skbio/sequence/_protein.py [hunk 3]
```diff
@@ -102,16 +103,19 @@ def _stop_codes(cls):
...
     def nondegenerate_chars(cls):
         return set("ACDEFGHIKLMNPQRSTVWY")
 
     @classproperty
     @overrides(IUPACSequence)
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_protein.py]: bump this on top of overrides

File: skbio/sequence/_rna.py [hunk 4]
```diff
@@ -98,6 +99,7 @@ class RNA(IUPACSequence, NucleotideMixin):
 
     @classproperty
     @overrides(NucleotideMixin)
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: bump this on top of overrides

File: skbio/sequence/_rna.py [hunk 5]
```diff
@@ -110,11 +112,13 @@ def complement_map(cls):
 
     @classproperty
     @overrides(IUPACSequence)
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: bump this on top of overrides

File: skbio/sequence/_rna.py [hunk 6]
```diff
@@ -110,11 +112,13 @@ def complement_map(cls):
...
     def nondegenerate_chars(cls):
         return set("ACGU")
 
     @classproperty
     @overrides(IUPACSequence)
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_rna.py]: bump this on top of overrides

File: skbio/sequence/_sequence.py [hunk 7]
```diff
@@ -373,6 +376,7 @@ def metadata(self):
         return self._metadata
 
     @metadata.setter
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_sequence.py]: I think the stability should apply to the property as a whole, @jairideout thoughts?

Comment (jairideout) [inline skbio/sequence/_sequence.py]: I agree, the getter/setter/deleter are linked to each other and we only have a docstring on the getter, so the getter should be the only one decorated.

Comment (gregcaporaso) [inline skbio/sequence/_sequence.py]: Ok, I made that change. It looks like this is only in one place in the code base right now - am I missing any?

Comment (jairideout) [inline skbio/sequence/_sequence.py]: From grepping (`grep -r '\.setter' *`) there is one more: `DissimilarityMatrix.ids`.

File: skbio/sequence/_sequence.py [hunk 8]
```diff
@@ -464,6 +468,7 @@ def positional_metadata(self):
         return self._positional_metadata
 
     @positional_metadata.setter
+    @stable(as_of="0.4.0")
```
Comment (jairideout) [inline skbio/sequence/_sequence.py]: remove

File: skbio/sequence/_sequence.py [hunk 9]
```diff
@@ -484,6 +489,7 @@ def positional_metadata(self, positional_metadata):
         self._positional_metadata = positional_metadata
 
     @positional_metadata.deleter
+    @stable(as_of="0.4.0")
```
Comment (jairideout) [inline skbio/sequence/_sequence.py]: remove

File: skbio/sequence/_sequence.py (hunk 7 above)
Comment (jairideout) [inline skbio/sequence/_sequence.py]: And there's `Sequence.positional_metadata` -- I just marked those on the diff.

File: skbio/sequence/_dna.py [hunk 10]
```diff
@@ -96,8 +97,9 @@ class DNA(IUPACSequence, NucleotideMixin):
 
     """
 
-    @classproperty
     @overrides(NucleotideMixin)
+    @stable(as_of="0.4.0")
+    @classproperty
```
Comment (jairideout) [inline skbio/sequence/_dna.py]: The correct ordering for this to work must be:

```python
@classproperty
@stable
@overrides
def foo():
    ...
```

Can you please update here and throughout?

Comment (gregcaporaso) [inline skbio/sequence/_dna.py]: Fixed, sorry, I mixed up the order in which these get applied.

File: skbio/sequence/_genetic_code.py [hunk 11]
```diff
@@ -179,6 +180,7 @@ def from_ncbi(cls, table_id=1):
                 % (sorted(_ncbi_genetic_codes), table_id))
         return _ncbi_genetic_codes[table_id]
 
+    @stable(as_of="0.4.0")
```
Comment (ebolyen) [inline skbio/sequence/_genetic_code.py]: classproperty on top

**Comment by ebolyen** [comment]
@gregcaporaso found one ordering issue. Could you also do a build of the docs locally to verify everything looks right?

**Comment by gregcaporaso** [comment]
@ebolyen, fixed the issue you noticed and spot-checked the docs fairly throughly (fixed a couple of issues).

**Comment by ebolyen** [comment]
Thanks @gregcaporaso

**Pull request merged**